packages/xstate-react/src/useSelector.ts

```diff
@@ -2,24 +2,41 @@ import { useCallback } from 'react';
 import { useSyncExternalStoreWithSelector } from 'use-sync-external-store/shim/with-selector';
 import { ActorRef, SnapshotFrom } from 'xstate';
 
+type SyncExternalStoreSubscribe = Parameters<
+  typeof useSyncExternalStoreWithSelector
+>[0];
+
 function defaultCompare<T>(a: T, b: T) {
   return a === b;
 }
 
 export function useSelector<TActor extends ActorRef<any, any>, T>(
```

Comment:
a non-overload version could look smth like this:
```diff
diff --git a/packages/xstate-react/src/createActorContext.ts b/packages/xstate-react/src/createActorContext.ts
index aa4668aa9..d217c2f6d 100644
--- a/packages/xstate-react/src/createActorContext.ts
+++ b/packages/xstate-react/src/createActorContext.ts
@@ -7,7 +7,8 @@ import {
   SnapshotFrom,
   ActorOptions,
   StateMachine,
-  AnyActorLogic
+  AnyActorLogic,
+  ActorRef
 } from 'xstate';
 
 export function createActorContext<TLogic extends AnyActorLogic>(
@@ -84,7 +85,7 @@ export function createActorContext<TLogic extends AnyActorLogic>(
     compare?: (a: T, b: T) => boolean
   ): T {
     const actor = useContext();
-    return useSelectorUnbound(actor, selector, compare);
+    return useSelectorUnbound(actor as ActorRef<any, any>, selector, compare);
   }
 
   return {
diff --git a/packages/xstate-react/src/useSelector.ts b/packages/xstate-react/src/useSelector.ts
index f2647d19f..96de665a2 100644
--- a/packages/xstate-react/src/useSelector.ts
+++ b/packages/xstate-react/src/useSelector.ts
@@ -10,19 +10,13 @@ function defaultCompare<T>(a: T, b: T) {
   return a === b;
 }
 
-export function useSelector<TActor extends ActorRef<any, any>, T>(
+export function useSelector<TActor extends ActorRef<any, any> | undefined, T>(
   actor: TActor,
-  selector: (emitted: SnapshotFrom<TActor>) => T,
-  compare?: (a: T, b: T) => boolean
-): T;
-export function useSelector<TActor extends ActorRef<any, any>, T>(
-  actor: TActor | undefined,
-  selector: (emitted: SnapshotFrom<TActor> | undefined) => T,
-  compare?: (a: T, b: T) => boolean
-): T;
-export function useSelector<TActor extends ActorRef<any, any>, T>(
-  actor: TActor | undefined,
-  selector: (emitted: SnapshotFrom<TActor> | undefined) => T,
+  selector: (
+    emitted: TActor extends ActorRef<any, any>
+      ? SnapshotFrom<TActor>
+      : undefined
+  ) => T,
   compare: (a: T, b: T) => boolean = defaultCompare
 ): T {
   const subscribe: SyncExternalStoreSubscribe = useCallback(
```

Reply:
I feel like the overload is more for helping devs; it's better to see a simpler type when they pass in an `actorRef` that won't be undefined.

Reply:
Yeah, that's why I didn't commit it. I just put it here for future reference. I don't think the more complex type is that big of an issue since TS displays resolved types~ with the inference result already substituted: [TS playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwxAGcMBGAHgBUA+AClgHMAueSgbQF0AaeRgJhbsOAShYA3HFmABuAFChIsBCnTY8BYmSp1GgzvAA+8NKERZUIYD34sTIMxeCj4EqXNmESpWm1Ij3CtBwSGiYuPieGHxU8CAAHoSowETGqADWqDgA7qj6RnYOlnSy8LwwzKxcJWUMAqmm5paxCSBJKZTwAPz19o3A8IJVzq4ysh6afD5+wnJAA). One could still see those in the source code though. Or when hovering over a referenced (but not called!) function, the conditional type version arguably displays better at that point. It might be slightly longer but at least it displays the whole thing instead of the first signature and the info that there is more there (`function test1<T>(arg: T[], arg2: T[]): void (+1 overload)`)

I'm still somewhat struggling when it comes to picking overloads over conditional types in situations like this. Overloads often suffer from cryptic errors but those probably happen in more complex scenarios. When it comes to adding just optionality to some arguments etc then overloads are quite fine from that PoV. I don't like how overloads don't quite work with some TS features (like conditional types, including "builtin" ones like `ReturnType` or `Parameters` but tbh they just don't work well with generic functions so 🤷‍♂️ ).

Reply:
I removed the overloads